File: christmas/programs/christmas/src/defs.rs
pub const DISCRIMINATOR_SIZE: usize = 8;
pub const PUBKEY_SIZE: usize = 32;
pub const U8_SIZE: usize = 1;
pub const U16_SIZE: usize = 2;
pub const U32_SIZE: usize = 4;
pub const U64_SIZE: usize = 8;
pub const BOOL_SIZE: usize = 1;
pub const STRING_PREFIX_SIZE: usize = 4;

pub const BUMP_SIZE: usize = 1;
pub const URI_SIZE: usize = STRING_PREFIX_SIZE + 204;
pub const GEO_SIZE: usize = STRING_PREFIX_SIZE + 6; // 6 characters of resolution
pub const TWO_FACTOR_SIZE: usize = U8_SIZE * 32; // 256 bit
pub const REGION_SIZE: usize = STRING_PREFIX_SIZE + 3; // 3 digit country code
pub const COUPON_NAME_SIZE: usize = STRING_PREFIX_SIZE + 36;
pub const STORE_NAME_SIZE: usize = STRING_PREFIX_SIZE + 36;
pub const DATE_SIZE: usize = U64_SIZE; // unix timestamp

pub const DATE_HASH_SIZE: usize = 32; // 32 * 8 = 256 days per epoch since 1 jan 2024
pub const DATE_HASH_OVERFLOW_SIZE: usize = BOOL_SIZE;
pub const HAS_SUPPLY_SIZE: usize = BOOL_SIZE;
pub const SUPPLY_SIZE: usize = U32_SIZE;
pub const DAYS_SINCE_1_JAN_2024: u64 = 19722; // Math.floor(new Date(2024, 0, 1).getTime() / MS_PER_DAY)
pub const MS_PER_DAY: u64 = 1000 * 60 * 60 * 24;
pub const DATE_HASH_BITS: u64 = DATE_HASH_SIZE as u64 * 8;

pub const REGION_CODES: &'static [&'static str] = &[
    "AFG", "ALB", "DZA", "ASM", "AND", "AGO", "AIA", "ATA", "ATG", "ARG", "ARM", "ABW", "AUS",
    "AUT", "AZE", "BHS", "BHR", "BGD", "BRB", "BLR", "BEL", "BLZ", "BEN", "BMU", "BTN", "BOL",
    "BES", "BIH", "BWA", "BVT", "BRA", "IOT", "BRN", "BGR", "BFA", "BDI", "CPV", "KHM", "CMR",
    "CAN", "CYM", "CAF", "TCD", "CHL", "CHN", "CXR", "CCK", "COL", "COM", "COD", "COG", "COK",
    "CRI", "HRV", "CUB", "CUW", "CYP", "CZE", "CIV", "DNK", "DJI", "DMA", "DOM", "ECU", "EGY",
    "SLV", "GNQ", "ERI", "EST", "SWZ", "ETH", "FLK", "FRO", "FJI", "FIN", "FRA", "GUF", "PYF",
    "ATF", "GAB", "GMB", "GEO", "DEU", "GHA", "GIB", "GRC", "GRL", "GRD", "GLP", "GUM", "GTM",
    "GGY", "GIN", "GNB", "GUY", "HTI", "HMD", "VAT", "HND", "HKG", "HUN", "ISL", "IND", "IDN",
    "IRN", "IRQ", "IRL", "IMN", "ISR", "ITA", "JAM", "JPN", "JEY", "JOR", "KAZ", "KEN", "KIR",
    "PRK", "KOR", "KWT", "KGZ", "LAO", "LVA", "LBN", "LSO", "LBR", "LBY", "LIE", "LTU", "LUX",
    "MAC", "MDG", "MWI", "MYS", "MDV", "MLI", "MLT", "MHL", "MTQ", "MRT", "MUS", "MYT", "MEX",
    "FSM", "MDA", "MCO", "MNG", "MNE", "MSR", "MAR", "MOZ", "MMR", "NAM", "NRU", "NPL", "NLD",
    "NCL", "NZL", "NIC", "NER", "NGA", "NIU", "NFK", "MNP", "NOR", "OMN", "PAK", "PLW", "PSE",
    "PAN", "PNG", "PRY", "PER", "PHL", "PCN", "POL", "PRT", "PRI", "QAT", "MKD", "ROU", "RUS",
    "RWA", "REU", "BLM", "SHN", "KNA", "LCA", "MAF", "SPM", "VCT", "WSM", "SMR", "STP", "SAU",
    "SEN", "SRB", "SYC", "SLE", "SGP", "SXM", "SVK", "SVN", "SLB", "SOM", "ZAF", "SGS", "SSD",
    "ESP", "LKA", "SDN", "SUR", "SJM", "SWE", "CHE", "SYR", "TWN", "TJK", "TZA", "THA", "TLS",
    "TGO", "TKL", "TON", "TTO", "TUN", "TUR", "TKM", "TCA", "TUV", "UGA", "UKR", "ARE", "GBR",
    "UMI", "USA", "URY", "UZB", "VUT", "VEN", "VNM", "VGB", "VIR", "WLF", "ESH", "YEM", "ZMB",
    "ZWE", "ALA",
];

File: christmas/app/src/lib/clients/anchor-client/defs.ts
export const PROGRAM_ID = "B2ejsK7m3eYPerru92hS73Gx7sQ7J83DKoLHGwn6pg5v";

export const DISCRIMINATOR_SIZE = 8;
export const PUBKEY_SIZE = 32;
export const U8_SIZE = 1;
export const U16_SIZE = 2;
export const U32_SIZE = 4;
export const U64_SIZE = 8;
export const BOOL_SIZE = 1;
export const STRING_PREFIX_SIZE = 4;
export const BUMP_SIZE = 1;
export const GEOHASH_SIZE = 6; // 6 characters of resolution
export const TWO_FACTOR_SIZE = U8_SIZE * 32; // 256 bit
export const REGION_SIZE = 3; // 3 digit country code
export const COUPON_NAME_SIZE = STRING_PREFIX_SIZE + 36;
export const COUPON_SYMBOL_SIZE = STRING_PREFIX_SIZE + 14;
export const STORE_NAME_SIZE = STRING_PREFIX_SIZE + 36;
export const URI_SIZE = STRING_PREFIX_SIZE + 204;
export const DATE_SIZE = U64_SIZE;
export const DATE_HASH_SIZE = 32; // 32 * 8 = 256 days per epoch since 1 jan 2024
export const DATE_HASH_OVERFLOW_SIZE = BOOL_SIZE;
export const HAS_SUPPLY_SIZE = BOOL_SIZE;
export const SUPPLY_SIZE = U32_SIZE;
export const DAYS_SINCE_1_JAN_2024 = 19722; // Math.floor(new Date(2024, 0, 1).getTime() / MS_PER_DAY)
export const MS_PER_DAY = 1000 * 60 * 60 * 24;
export const DATE_HASH_BITS = DATE_HASH_SIZE * 8;

export const USER_ACCOUNT_SIZE =
    DISCRIMINATOR_SIZE + REGION_SIZE + GEOHASH_SIZE + BUMP_SIZE;

export const COUPON_ACCOUNT_SIZE =
    DISCRIMINATOR_SIZE +
    PUBKEY_SIZE + // update_authority
    PUBKEY_SIZE + // mint
    COUPON_NAME_SIZE +
    URI_SIZE +
    // convenience fields
    PUBKEY_SIZE + // store
    DATE_SIZE + // valid_from
    DATE_SIZE + // valid_to
    SUPPLY_SIZE +
    // memcmp block
    REGION_SIZE +
    GEOHASH_SIZE +
    HAS_SUPPLY_SIZE +
    DATE_HASH_SIZE + // valid_from_hash
    DATE_HASH_SIZE + // valid_to_hash
    DATE_HASH_OVERFLOW_SIZE +
    // bump
    BUMP_SIZE;

export const OFFSET_TO_STORE =
    DISCRIMINATOR_SIZE +
    PUBKEY_SIZE + // update_authority
    PUBKEY_SIZE + // mint
    COUPON_NAME_SIZE +
    URI_SIZE;

export const OFFSET_TO_MEMCMP_BLOCK =
    DISCRIMINATOR_SIZE +
    PUBKEY_SIZE + // update_authority
    PUBKEY_SIZE + // mint
    COUPON_NAME_SIZE +
    URI_SIZE +
    // convenience fields
    PUBKEY_SIZE + // store
    DATE_SIZE + // valid_from
    DATE_SIZE + // valid_to
    SUPPLY_SIZE;

export const OFFSET_TO_VALID_FROM_HASH =
    DISCRIMINATOR_SIZE +
    PUBKEY_SIZE + // update_authority
    PUBKEY_SIZE + // mint
    COUPON_NAME_SIZE +
    URI_SIZE +
    // convenience fields
    PUBKEY_SIZE + // store
    DATE_SIZE + // valid_from
    DATE_SIZE + // valid_to
    SUPPLY_SIZE +
    // memcmp block
    REGION_SIZE +
    GEOHASH_SIZE +
    HAS_SUPPLY_SIZE;

export const OFFSET_TO_VALID_TO_HASH =
    DISCRIMINATOR_SIZE +
    PUBKEY_SIZE + // update_authority
    PUBKEY_SIZE + // mint
    COUPON_NAME_SIZE +
    URI_SIZE +
    // convenience fields
    PUBKEY_SIZE + // store
    DATE_SIZE + // valid_from
    DATE_SIZE + // valid_to
    SUPPLY_SIZE +
    // memcmp block
    REGION_SIZE +
    GEOHASH_SIZE +
    HAS_SUPPLY_SIZE +
    DATE_HASH_SIZE; // valid_from_hash

export const OFFSET_TO_REGION =
    DISCRIMINATOR_SIZE +
    PUBKEY_SIZE + // update_authority
    PUBKEY_SIZE + // mint
    COUPON_NAME_SIZE +
    URI_SIZE +
    // convenience fields
    PUBKEY_SIZE + // store
    DATE_SIZE + // valid_from
    DATE_SIZE + // valid_to
    SUPPLY_SIZE;

export const OFFSET_TO_HAS_SUPPLY =
    DISCRIMINATOR_SIZE +
    PUBKEY_SIZE + // update_authority
    PUBKEY_SIZE + // mint
    COUPON_NAME_SIZE +
    URI_SIZE +
    // convenience fields
    PUBKEY_SIZE + // store
    DATE_SIZE + // valid_from
    DATE_SIZE + // valid_to
    SUPPLY_SIZE +
    // memcmp block
    REGION_SIZE +
    GEOHASH_SIZE;

export const OFFSET_TO_DATE_HASH_OVERFLOW =
    DISCRIMINATOR_SIZE +
    PUBKEY_SIZE + // update_authority
    PUBKEY_SIZE + // mint
    COUPON_NAME_SIZE +
    URI_SIZE +
    // convenience fields
    PUBKEY_SIZE + // store
    DATE_SIZE + // valid_from
    DATE_SIZE + // valid_to
    SUPPLY_SIZE +
    // memcmp block
    REGION_SIZE +
    GEOHASH_SIZE +
    HAS_SUPPLY_SIZE +
    DATE_HASH_SIZE + // valid_from_hash
    DATE_HASH_SIZE;

export const OFFSET_TO_GEO =
    DISCRIMINATOR_SIZE +
    PUBKEY_SIZE + // update_authority
    PUBKEY_SIZE + // mint
    COUPON_NAME_SIZE +
    URI_SIZE +
    // convenience fields
    PUBKEY_SIZE + // store
    DATE_SIZE + // valid_from
    DATE_SIZE + // valid_to
    SUPPLY_SIZE +
    // memcmp block
    REGION_SIZE;
